### PR TITLE
Issue 996 - Basic reports

### LINKED ISSUE #996: Implement basic reporting functionality for MVP (besides Domain Growth report)
### Story

As an Admin, I want to quickly export domain data reports directly from the domains page (/admin/registrar/domain/) so that I can easily access and analyze the domain data.

### Acceptance Criteria

- [ ] Three reports are available to download on the domains page:
  - [ ] [Domains by type](https://docs.google.com/spreadsheets/d/1_nMU2obW22U6NlOSC2ARxf3PpsJnSe2wMo5AyLSzXzk/edit?usp=sharing) (sorted by domain name)
  - [ ] [current-full.csv](https://github.com/cisagov/dotgov-data/blob/main/current-full.csv) (sorted by domain name, then agency, then domain type)
  - [ ] [current-federal.csv](https://github.com/cisagov/dotgov-data/blob/main/current-federal.csv) (sorted by domain name, then agency, then domain type)
- [ ] Each CSV report should contain accurate and up-to-date domain data from the database, sorted in the ways they are in the examples above.
- [ ] Single dropdown with the three report options which the user can select
- [ ] Clicking on each report triggers an immediate download of the relevant CSV report
- [ ] The UI components should be consistent with the existing design language of the admin portal.

### Additional Context

- This feature is a stop-gap measure, meant to provide immediate access to crucial reports while the ideal report interface is being developed. Future work is at #997.

- Security email may be pulled from .gov database rather than thru EPP call to registry.

### Issue Links

🔄 Relates to: #938 #143 #1075

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #996 

## Changes

- Override DomainAdmin's change_list with a template that has buttons
- Link buttons to actions in admin to trigger query and export methods
- Query and export methods for the 3 reports
- Unit test for the core functinality
- Fixed CSS on those object-tools buttons

## Context for reviewers

Note that we're querying DomainInformation, so if a Domain does not have a DI (transition domain?), that domain will be excluded. We can refactor to query on Domains instead, if needed.

## Code Review Verification Steps

You will need READY domains to see data exported, so you may need to hack the domains or have the filter look for UNKNOWN domains instead (lines 60, 77, 96 in csv_export.py)

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [x] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [x] Pulled this branch locally and tested it
- [x] Reviewed this code and left comments
- [x] Checked that all code is adequately covered by tests
- [x] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [x] All new functions and methods are commented using plain language
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

Screen Shot 2023-10-31 at 1 51 04 PM

Screen Shot 2023-10-31 at 1 50 54 PM

Screen Shot 2023-10-31 at 1 50 48 PM

### CODE REVIEW

**Review by zandercymatics — APPROVED** [review]
Great work! I did notice some styling weirdness at 150% zoom (I'm using a 1080p monitor on my laptop) which I think can be fixed with something like min-width (.minw). Bordering on the optional territory. Looks good otherwise. Won't block on it though as the main code changes are minimal
image


Tested with ANDI as well and it seems to be fine. I did notice that its yelling about some tab index but I am not sure if its actually related to your change. Additionally, the other buttons don't have title attributes but neither does the "add domain" button so I think you are fine there as well.
image
image

File: src/registrar/admin.py [hunk 1]
```diff
@@ -747,8 +749,57 @@ def organization_type(self, obj):
...
+                "export_data_full/",
+                self.export_data_full,
+                name="%s_%s_export_data_full" % info,
+            ),
+            path(
+                "export_data_federal/",
+                self.export_data_federal,
+                name="%s_%s_export_data_federal" % info,
+            ),
+        ]
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/admin.py:799]: Yep - we're pulling READY, ON_HOLD and DNS_NEEDED domains. There will NEVER be a case where there are no domains that qualify in our production app.

File: src/registrar/tests/test_reports.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+        # Normalize line endings and remove commas,
+        # spaces and leading/trailing whitespace
+        csv_content = (
+            csv_content.replace(",,", "")
+            .replace(",", "")
+            .replace(" ", "")
+            .replace("\r\n", "\n")
+            .strip()
+        )
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/tests/test_reports.py:130]: It outputs weird commas when there are no values (empty strings) and drove me nuts in testing, so I went nuclear with the sanitizing. I guess I should be able to leave commas alone seeing as I'm replacing whitespaces.

File: src/registrar/tests/test_views.py [hunk 3]
```diff
@@ -1170,6 +1170,7 @@ def tearDown(self):
             if hasattr(self.domain, "contacts"):
                 self.domain.contacts.all().delete()
             DomainApplication.objects.all().delete()
+            DomainInformation.objects.all().delete()
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/tests/test_views.py:1173]: Yep, saw that it was a missing piece of cleanup.

File: src/registrar/utility/csv_export.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            "Domain type": domainInfo.organization_type,
+            "Federal agency": domainInfo.federal_agency,
+            "Organization name": domainInfo.organization_name,
+            "City": domainInfo.city,
+            "State": domainInfo.state_territory,
+            "AO": domainInfo.authorizing_official.first_name
+            + " "
+            + domainInfo.authorizing_official.last_name
+            if domainInfo.authorizing_official
+            else " ",
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/utility/csv_export.py:35]: Nature of the data

File: src/registrar/tests/test_reports.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+                Domain.State.DNS_NEEDED,
+                Domain.State.ON_HOLD,
+            ],
+        }
+
+        # Call the export function
+        export_domains_to_writer(writer, columns, sort_fields, filter_condition)
+
+        # Reset the CSV file's position to the beginning
+        csv_file.seek(0)
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/tests/test_reports.py:106]: Nope.

[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/tests/test_reports.py:106]: It's defaults to the end, I think.

File: src/registrar/templates/django/admin/domain_change_list.html [hunk 6]
```diff
@@ -0,0 +1,23 @@
+{% extends "admin/change_list.html" %}
+
+{% block object-tools %}
+
+<ul class="object-tools">
+    <li>
+        <a href="{% url 'admin:registrar_domain_export_data_type' %}" class="button">Export all domain metadata</a>
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/templates/django/admin/domain_change_list.html:7]: Going by @h-m-f-t's suggested copy. Easily revised.

File: src/registrar/tests/test_reports.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/tests/test_reports.py:130]: Nope, the commas still trip me up.

File: src/registrar/templates/django/admin/domain_change_list.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (zandercymatics) [reply src/registrar/templates/django/admin/domain_change_list.html:7]: Gotcha

File: src/registrar/tests/test_reports.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zandercymatics) [reply src/registrar/tests/test_reports.py:130]: Yeah the weird commas are a pain to look at. I've seen stuff like ",,,,,data,,,," before. I personally think its OK as long as the test works consistently and predictably. That said @abroddrick may have an alternate opinion here

File: src/registrar/admin.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zandercymatics) [reply src/registrar/admin.py:799]: Fair enough then! Marking as resolved

File: src/registrar/tests/test_reports.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zandercymatics) [reply src/registrar/tests/test_reports.py:106]: Ah, thats right. It reads bottom to top. Thanks

File: src/registrar/assets/sass/_theme/_admin.scss [hunk 7]
```diff
@@ -180,3 +180,48 @@ h1, h2, h3 {
...
+#changelist {
+    clear: both;
+}
+
+// Account for the h2, roughly 90px
+@include at-media(tablet) {
+    .object-tools {
+        padding-left: 90px;
+    }
+}
```
Comment (abroddrick) [inline src/registrar/assets/sass/_theme/_admin.scss:227]: end of file line missing

File: src/registrar/templates/django/admin/domain_change_list.html [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    </li>
+    {% if has_add_permission %}
+    <li>
+        <a href="{% url 'admin:registrar_domain_add' %}" class="addlink">
+            Add Domain
+        </a>
+    </li>
+    {% endif %}
+</ul>
+{% endblock %}
```
Comment (abroddrick) [inline src/registrar/templates/django/admin/domain_change_list.html:23]: EOF line missing here too

File: src/registrar/tests/test_reports.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (abroddrick) [reply src/registrar/tests/test_reports.py:130]: It stills works to download with removing the extra commas? I get doing that for visual display as long as it doesn't affect cameron's ability to post it via github scrapping or affect a users ability to download it locally

File: src/registrar/admin.py [hunk 9]
```diff
@@ -747,8 +749,59 @@ def organization_type(self, obj):
     search_fields = ["name"]
     search_help_text = "Search by domain name."
     change_form_template = "django/admin/domain_change_form.html"
+    change_list_template = "django/admin/domain_change_list.html"
     readonly_fields = ["state"]
 
+    def export_data_type(self, request):
```
Comment (therealslimhsiehdy) [inline src/registrar/admin.py:755]: Q: Would it be worth leaving a comment on what each CSV is for + how they're sorted? Dunno if that would be useful

ie `domains-by-type - domains in all on hold/dns needed/ready status sorted by domain name`

**Review by therealslimhsiehdy — COMMENTED** [review]
Let a smol comment, address if you see you see fit! Otherwise this looks pretty great!

**Comment by Katherine-Osos** [comment]
Looks good to me! 

Non-blocking comment: Could we use color to differentiate the report buttons vs the "add domain" button? e.g. the report buttons could be white with black lettering while the "add domain" remains as is? Can also think thru this in a future iteration.

**Review by Katherine-Osos — APPROVED** [review]

**Pull request merged**